### ISSUE TITLE
Add a local PlifePred2 prediction adapter after resolving output semantics and platform support

### ISSUE THREAD

**Issue by iskandr** [openvax/mhctools#283]
## Request

Evaluate and add the downloadable PlifePred2 prediction interface as a concrete child of #279. This is a newer package, not an assertion that the original 2018 PlifePred web implementation is locally available or scientifically equivalent.

## Artifact evidence (2026-09-09)

- Published package: https://pypi.org/project/plifepred2/1.0/.
- Source: https://github.com/shindebpratik/plifepred2 (the older patrik-ackerman URL redirects here), revision `5e4e0c6ab6eea63f479316df60985d64c5468c7e`.
- Inspected the published wheel in memory and verified its SHA-256: `e0eb32928b7970d5a113d8826ef9e451c94f0156e0b628dddb0da7103ac2c32e`.
- Wheel contains the prediction CLI, two trained `.sav` models, property resources, `pfeature_comp` executable and a GPLv3 LICENSE.
- Requires Python >=3.10 and scikit-learn 1.4.2. The wheel is tagged `py3-none-any` despite bundling an executable; determine actual supported operating systems/architectures rather than trusting the tag.
- The README's raghavagps clone URL returned 404 during this audit. Use the verified repository/package instead.
- No third-party executable or model was run during the availability audit.

## Important semantics gate

The README describes a random-forest regressor but calls the output `Halflife` a probability; prediction code can use either `predict_proba` or `predict`. Establish the actual trained model type, target definition, units/transform and supporting publication before exposing a serum-half-life field. Do not inherit ambiguous upstream labels.

## Acceptance

- [ ] Confirm scientific provenance and whether each model estimates a duration, transformed target, or classification score. Distinguish serum/plasma/blood and in-vivo measurements.
- [ ] Reproduce inference with pinned package/model versions and reviewed artifact loading. Review GPL obligations and keep integration optional and appropriately isolated.
- [ ] Establish executable architecture, runtime requirements and executable permissions; unsupported platforms must fail clearly. Do not downgrade the host mhctools environment.
- [ ] Implement prediction-only support. Do not wrap the mutation-design interface.
- [ ] Preserve each input ID and explicit rejection reason; upstream length/alphabet filtering must not silently drop or misalign rows.
- [ ] Handle natural and modified-input modes explicitly; coarse modification flags are not a substitute for exact chemical-form provenance.
- [ ] Add deterministic parser tests and an opt-in local smoke test with no network during inference. Verify units/labels against documented reference outputs and inspect missing-executable/model failure behavior.
- [ ] Keep this endpoint separate from protease cut-site scores and pMHC stability.

Downstream: openvax/topiary#288 and Vaxrank's construct audit.

**Comment by iskandr** [comment]
Additional platform verification: the published `pfeature_comp` release asset at https://github.com/shindebpratik/plifepred2/releases/download/v.0.1/pfeature_comp begins with an ELF header, class 2, machine 62 (x86-64). It is not a native macOS executable. This was a 64-byte header inspection, not execution. Native Apple Silicon support must not be inferred from the README or the wheel's py3-none-any tag; document a validated compatible runtime or obtain a supported feature extractor. The wheel contains an executable of the same published size, but its own architecture and runtime still belong in the installation smoke test.

**Comment by iskandr** [comment]
## Executed verification (2026-09-09)

The published wheel was downloaded, its hash re-checked, and its two models loaded and introspected. The bundled executable was **not** run.

### The semantics gate is resolved: the output is not a probability

`plifepred2-1.0-py3-none-any.whl`, sha256 `e0eb32928b7970d5a113d8826ef9e451c94f0156e0b628dddb0da7103ac2c32e` (matches PyPI).

Both shipped models are **`sklearn.ensemble._forest.RandomForestRegressor`**, `criterion="squared_error"`, 100 estimators, `_sklearn_version` 1.4.2:

| model | file | `n_features_in_` |
| --- | --- | --- |
| Model 1 (natural) | `models/plifepred2_natural_model.sav` | 42 |
| Model 2 (modified) | `models/plifepred2_model.sav` | 105 |

The CLI does `if hasattr(model, "predict_proba"): prob = model.predict_proba(X)[:, 1] else: prob = model.predict(X)`. A `RandomForestRegressor` has no `predict_proba`, so **the branch is dead and `model.predict` is always what runs**. The `METADATA` table describing the `Halflife` column as "Predicted probability" is simply wrong, and so is the local variable name `prob`. The README's own "Machine Learning Model / Algorithm: Random Forest Regressor" section is the accurate one.

### What the target actually is

From the fitted trees (leaf values across estimators, and each tree's root value = mean training target):

| model | leaf value range | root mean across trees |
| --- | --- | --- |
| natural | 1.342 – 5.782 | 3.294 |
| modified | 1.305 – 6.914 | 3.835 |

A regression target bounded in roughly 1.3–6.9 is a **log-transformed duration**, not a probability and not a duration in raw units. It is consistent with `log10(half-life in seconds)` (10^1.3 ≈ 20 s to 10^6.9 ≈ 100 d) but `ln(minutes)` is not excluded, and the package cites no publication ("please cite the corresponding publication", with no reference given). **Base and unit remain unresolved**, so the first acceptance item is only partly closed: we now know it is a duration on a log scale, not which one.

Practical consequence: this must not be exposed in a units-bearing field until the transform is pinned down. Since the target is monotone in half-life, it is usable as a rank/score in the meantime.

### Platform

`plifepred2/pfeature_comp` is:

```
ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked,
interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, stripped
```

45 MB, and the wheel is nonetheless tagged `py3-none-any`. **Rosetta translates x86-64 macOS binaries, not Linux ELF**, so there is no native or emulated macOS path — it needs Docker or a Linux VM. The METADATA claim that the toolkit "is optimized for Linux and macOS environments" does not hold for the bundled extractor.

### There is a way around the binary for model 1

`feature_names_in_` on the natural model is exactly 42 features:

```
QSO1_SC_A … QSO1_SC_Y   (20, Schneider-Wrede distance matrix)
QSO1_G_A  … QSO1_G_Y    (20, Grantham distance matrix)
QSO1_SC1, QSO1_G1       (2, lag-1 coupling terms)
```

That is Pfeature's quasi-sequence-order descriptor at `nlag=1` and nothing else — a small, published descriptor that can be reimplemented clean-room and validated against `pfeature_comp` output, the same approach already used for ERAMER. Model 2 needs 105 features spanning AAC, DPC, PCP, RRI, DDR, SEP, SER, SPC, CTC, CeTD, PAAC, APAAC and QSO, which is not worth reimplementing. So a **natural-peptide-only adapter can run on macOS**; modified-peptide support stays gated on the Linux binary.

### Adapter notes found while reading the CLI

- Upstream filtering rejects anything outside 12–100 residues or containing non-standard residues, and writes those rows to **`eliminated_sequences.csv` in the current working directory** — a hardcoded relative path. It also writes `<tmpname>_features.csv` next to the output file. Any adapter must run this in a scratch cwd and read the rejection file back, which is how the "preserve each input ID and explicit rejection reason" item gets satisfied.
- Modification flags for model 2 are set as five all-or-nothing columns (`d`, `ct`, `nt`, `cyc`, `ptm`) applied uniformly to **every** sequence in the input file, not per sequence. Batching peptides with different modifications through one call would mislabel them.
- The `required_columns` list carries a stale comment ("keep your full list here unchanged") and a `CeTD` column-rename fix that patches up positional column names `"2"`, `"2.2"`, `"3.2"` coming out of the extractor — a sign the feature matrix column naming is fragile across extractor versions. Pin the extractor.

### Licensing

Wheel ships a full GPLv3 `LICENSE`, and `METADATA` declares `License: GPLv3`. That points at a subprocess/sidecar integration rather than an in-process import, which the `scikit-learn==1.4.2` pin forces anyway.

### Kind

Same conclusion as #280: needs a new allele-independent `serum_half_life` kind. PlifePred2 would populate `score` only, with no `value`, until the log base and unit are established.

**Comment by iskandr** [comment]
## Additional gate: the matrix is blood, not serum

Checked while designing the prediction kind for this endpoint, and it changes where PlifePred2 can land.

The lineage paper (PMC5983457, PLoS ONE 2018) is titled **"In silico approaches for predicting the half-life of natural and modified peptides in blood"**, and its abstract specifies "163 natural and 98 modified peptides whose half-life has been determined experimentally in **mammalian blood**" — 261 total, matching the count in #279. So the original PlifePred is a *mammalian blood* model, not a human *serum* model.

Two consequences:

1. **It cannot share a field with PeptiVerse (#280)**, whose half-life endpoint is explicitly human serum in hours. #279's requirement to keep serum, plasma and whole blood distinct applies directly here, and nothing in an `mhctools.pred.Prediction` records the assay matrix — the kind string is the only place that distinction can live. The `serum_half_life` kind being added for #280 would therefore be a mislabel for this predictor.
2. **PlifePred2's own matrix is unestablished.** The package cites no publication ("please cite the corresponding publication", with no reference), and its descriptors differ from the 2018 model's — PlifePred2 uses Pfeature QSO/composition features where the 2018 paper used PaDEL chemical descriptors and reported r = 0.643 (natural, composition) to 0.743 (natural, PaDEL). So it is a different model family, and inheriting the 2018 paper's matrix by assumption is exactly the kind of label inheritance this issue warns against.

Adding an acceptance item ahead of the platform work: **establish PlifePred2's actual training set, matrix and species from the authors or a publication** before choosing its kind. If it is blood, it needs a `blood_half_life` kind of its own rather than reuse of `serum_half_life`; if it turns out to be mammalian rather than human, that is a further caveat for downstream ranking, since the species is likewise unrecorded in a `Prediction`.

This does not block the clean-room QSO work — a 42-feature `nlag=1` QSO extractor is worth having regardless of how the endpoint is finally labelled — but it does block exposing the output under any named half-life field.

**Comment by iskandr** [comment]
## The Linux-binary blocker is removable, and clean-room would have been wrong

Two findings from actually running the model. **PlifePred2 model 1 now runs end-to-end on macOS arm64, with no Linux binary and no container.**

### Use Pfeature's Python source, not the frozen binary

The bundled `pfeature_comp` is a **PyInstaller onefile** (contains `_MEIPASS`, bundles `libpython3.9.so.1.0`), frozen from `raghavagps/Pfeature`'s `Standalone/pfeature_comp.py` — which is plain, portable Python. That source runs natively here and emits **exactly** the 42 columns model 1 declares in `feature_names_in_`, in the same order:

```
$ python3 pfeature_comp.py -i in.fa -o qso_out.csv -j QSO
Order of Gap: 1  ;Value of weight: 0.1
```

```python
X = pd.read_csv("qso_out.csv")
m = joblib.load("plifepred2_natural_model.sav")
list(X.columns) == list(m.feature_names_in_)   # True
m.predict(X)                                   # array([3.1572664, 3.79347232])
```

So the platform acceptance item can be met by wrapping upstream's own extractor rather than the frozen build. Notes for whoever implements it:

- `qos()` reads `Data/Schneider-Wrede.csv` and `Data/Grantham.csv` by **relative path**, so it must run with `cwd` set to the Pfeature `Standalone` directory — the same hardcoded-relative-path pattern already handled for DeepImmuno and TLimmuno2.
- Pfeature is **GPLv3**, like PlifePred2 itself, so this stays wrap-don't-vendor: a subprocess against a user-provided checkout, nothing vendored.
- The last change to `Standalone/pfeature_comp.py` upstream was 2024-06-07 and the wheel's frozen binary is dated 2024-09-18, so the source is very likely the same version the binary was built from. Confirming that against the binary under Linux remains open, but is no longer on the critical path.

### A clean-room QSO reimplementation would have been silently wrong

Retracting the suggestion in my earlier comment that the 42 QSO features be reimplemented clean-room. Pfeature's `qos()` has a copy-paste bug:

```python
for k in range(0,gap):
    c3.append((w * zz[k][i])  / (1 + w * zz['sum'][i]))   # Schneider-Wrede
    c4.append((w * zz[k][i])  / (1 + w * zz['sum'][i]))   # should be zz2 (Grantham)
```

`c4` reads `zz` where it should read `zz2`, so **`QSO1_G1` is a byte-for-byte duplicate of `QSO1_SC1`** rather than the Grantham lag-1 coupling term. Confirmed on real output:

```
QSO1_SC1=0.3389, QSO1_G1=0.3389
QSO1_SC1=0.2995, QSO1_G1=0.2995
```

Model 1 was trained on features carrying this bug, so reproducing its predictions requires reproducing the bug. An implementation written from the published quasi-sequence-order definition would put a genuine Grantham value in that column and produce silently wrong half-lives — no error, just different numbers. Wrapping upstream's extractor avoids the whole class of problem.

### Units still unresolved

With the pipeline running, the ambiguity from the previous comment is unchanged. `m.predict` returns 3.157 and 3.793 for two test peptides, which is ~24 min as `log10(seconds)` and ~23.5 min as `ln(minutes)` — both plausible, and the coincidence in this range means the output alone cannot settle it. Still needs the authors or a publication, and combines with the matrix question above: the endpoint cannot be exposed under a named half-life field until both are answered, though it is usable as a rank-only score.

**Comment by iskandr** [comment]
## Units resolved: log10 of half-life in seconds

The semantics gate is now closed. The transform can be read directly off the fitted trees, and the answer is unambiguous.

Mathur et al. 2018 (PMC5983457) documents the dataset construction: peptides drawn from PEPlife, filtered to those "experimentally validated in **mammalian blood**", then "we removed all those peptides having half-life more than 24 hours and **less than 20 seconds**", and "in order to present half-life on a linear scale we have taken **log2** of the half-life of peptides in seconds."

That gives a testable prediction for each candidate transform. Taking the min and max leaf value over all 100 trees in each PlifePred2 forest — the extremes of the training targets, since a leaf value is the mean of the targets it holds:

| model | min leaf | max leaf | as log10 s | as log2 s | as ln s |
|---|---|---|---|---|---|
| natural | 1.34242 | 5.78161 | 22.0 s → **604800.0 s** | 2.5 s → 55 s | 3.8 s → 324 s |
| modified | 1.30535 | 6.91424 | **20.2 s** → 8208000.0 s | 2.5 s → 121 s | 3.7 s → 1007 s |

Only the log10-seconds reading is coherent:

- The modified model's minimum is **20.2 s**, reproducing the paper's documented 20-second floor (log10 20 = 1.30103 vs. an observed 1.30535).
- The natural model's maximum inverts to **604800.0 seconds — exactly 7.000 days**. The modified model's maximum is **8208000 seconds — exactly 95.0 days**. Round durations like that are real measurements, not artifacts of a wrong transform.
- log2 and ln both put the entire training range below or around the documented 20-second floor, which is impossible.

So **PlifePred2's target is `log10(half-life in seconds)`**, and to recover a duration: `seconds = 10 ** prediction`.

Two ways PlifePred2 departs from the 2018 paper, which reinforce that it is a distinct model rather than a repackaging:

1. **The transform changed from log2 to log10.** Anyone porting assumptions from the paper will be wrong by a constant factor of log2(10) ≈ 3.32.
2. **The 24-hour cap is gone.** The paper capped at 24 h (log10 = 4.937); these forests hold targets out to 7 days and 95 days. Combined with the CLI's 12–100 residue filter versus the paper's 5–50, PlifePred2 was clearly trained on a larger, later dataset — plausibly the PEPlife2 resource tracked in #286, which is worth checking for train/test overlap before either is treated as independent validation of the other.

## Kind

With units and matrix both settled, the endpoint is: **ex vivo degradation half-life of a free peptide incubated in mammalian whole blood**. That is neither `pMHC_stability` nor the `serum_half_life` kind added for PeptiVerse in #280 — whole blood is not serum, and the species is mammalian-mixed rather than human. It needs its own `blood_half_life` kind, with `value` in hours derived as `10 ** score / 3600`.

Remaining caveat to carry into any adapter: the natural-peptide model is the weaker of the pair in the lineage paper (r = 0.643 for composition features versus 0.743 for chemical descriptors), and 90%-similar sequences were deliberately retained in the training data, so cross-validation numbers are optimistic for novel sequences.
